source/common/network/listener_impl.cc

```diff
@@ -9,19 +9,53 @@
 #include "common/network/connection_impl.h"
 #include "common/ssl/connection_impl.h"
 
+#include "server/connection_handler.h"
+
 #include "event2/listener.h"
 
 namespace Network {
 
+void ListenerImpl::listenCallback(evconnlistener*, evutil_socket_t fd, sockaddr* addr, int,
+                                  void* arg) {
+  ListenerImpl* listener = static_cast<ListenerImpl*>(arg);
+
+  if (listener->use_original_dst_ && listener->connection_handler_ != nullptr) {
+    struct sockaddr_storage orig_dst_addr;
+    memset(&orig_dst_addr, 0, sizeof(orig_dst_addr));
+
+    bool success = Utility::getOriginalDst(fd, &orig_dst_addr);
+
+    if (success) {
+      std::string orig_sock_name = std::to_string(
+          listener->getAddressPort(reinterpret_cast<struct sockaddr*>(&orig_dst_addr)));
+
+      if (listener->socket_.name() != orig_sock_name) {
+        ListenerImpl* new_listener =
+            static_cast<ListenerImpl*>(listener->connection_handler_->findListener(orig_sock_name));
+
+        if (new_listener != nullptr) {
+          listener = new_listener;
+        }
+      }
+    }
+  }
+
+  listener->newConnection(fd, addr);
+}
+
 ListenerImpl::ListenerImpl(Event::DispatcherImpl& dispatcher, ListenSocket& socket,
-                           ListenerCallbacks& cb, Stats::Store& stats_store, bool use_proxy_proto)
-    : dispatcher_(dispatcher), cb_(cb), use_proxy_proto_(use_proxy_proto),
-      proxy_protocol_(stats_store) {
-  listener_.reset(
-      evconnlistener_new(&dispatcher_.base(),
-                         [](evconnlistener*, evutil_socket_t fd, sockaddr* addr, int, void* arg)
-                             -> void { static_cast<ListenerImpl*>(arg)->newConnection(fd, addr); },
-                         this, 0, -1, socket.fd()));
+                           ListenerCallbacks& cb, Stats::Store& stats_store, bool bind_to_port,
+                           bool use_proxy_proto, bool use_orig_dst)
+    : dispatcher_(dispatcher), socket_(socket), cb_(cb), bind_to_port_(bind_to_port),
+      use_proxy_proto_(use_proxy_proto), proxy_protocol_(stats_store),
+      use_original_dst_(use_orig_dst), connection_handler_(nullptr) {
+
+  if (bind_to_port_) {
+    listener_.reset(
+        evconnlistener_new(&dispatcher_.base(), listenCallback, this, 0, -1, socket.fd()));
+  } else {
+    listener_.reset(evconnlistener_new(&dispatcher_.base(), nullptr, this, 0, -1, socket.fd()));
```

Comment:
can we ASSERT(-1 == socket.fd()) here or even have a dummy callback that panics? I guess do we even need a listener_ in this case? I'm not sure why we would. Leaving as nullptr would be better I think.